### PR TITLE
[simulation] simulate energy scan

### PR DESCRIPTION
**Pull request by bukepo**
This commit implements energy scan in simulation platform.

Verified locally
```bash
> ifconfig up
Done
> scan energy 100
| Ch | RSSI |
+----+------+
| 11 |  -98 |
| 12 |  -98 |
| 13 |  -98 |
| 14 |  -98 |
| 15 |  -98 |
| 16 |  -98 |
| 17 |  -30 |
| 18 |  -98 |
| 19 |  -98 |
| 20 |  -30 |
| 21 |  -30 |
| 22 |  -98 |
| 23 |  -30 |
| 24 |  -98 |
| 25 |  -30 |
| 26 |  -30 |
Done
```

### CODE REVIEW

File: examples/platforms/simulation/system.c [hunk 1]
```diff
@@ -127,6 +130,9 @@ void otSysInit(int aArgCount, char *aArgVector[])
         case OT_SIM_OPT_HELP:
             PrintUsage(aArgVector[0], EXIT_SUCCESS);
             break;
+        case OT_SIM_OPT_ENABLE_ENERGY_SCAN:
+            gRadioCaps |= OT_RADIO_CAPS_ENERGY_SCAN;
```
Comment (abtink) [inline examples/platforms/simulation/system.c:134]: Hi Yakun, what is the background/use-case of adding energy scan support in radio platform? Why do we need to have energy scan in simulation?

One of the key goals of simulation platform is to verify and validate the implementation in OT core. So it make sense to not implement and include radio capabilities in simulation radio itself and leave it to OT core implementation to be used (so we validate the implementation in core in `SubMac`/`Mac`).

`scan energy` does work now and it uses the implementation in `SubMac` along with `otPlatRadioGetRssi()`. If you want to make it look more random, maybe can consider changing the `otPlatRadioGetRssi()` code:

```c++
int8_t otPlatRadioGetRssi(otInstance *aInstance)
{
    OT_UNUSED_VARIABLE(aInstance);

    assert(aInstance != NULL);

    int8_t   rssi    = SIM_LOW_RSSI_SAMPLE;
    uint8_t  channel = sReceiveFrame.mChannel;
    uint32_t probabilityThreshold;

    otEXPECT((SIM_RADIO_CHANNEL_MIN <= channel) && channel <= (SIM_RADIO_CHANNEL_MAX));

    // To emulate a simple interference model, we return either a high or
    // a low  RSSI value with a fixed probability per each channel. The
    // probability is increased per channel by a constant.

    probabilityThreshold = (channel - SIM_RADIO_CHANNEL_MIN) * SIM_HIGH_RSSI_PROB_INC_PER_CHANNEL;

    if (otRandomNonCryptoGetUint16() < (probabilityThreshold * 0xffff / 100))
    {
        rssi = SIM_HIGH_RSSI_SAMPLE;
    }

exit:
    return rssi;
}
```

This code now returns either a high (-30) or low (-98) value randomly based on the channel.
- There is pattern which ensures on average lower channel numbers look better (less noisy) which is what is assumed/relied upon in some of the test cases around "channel monitor" and "channel manager" (which use simulation platform).
- We can add randomness (pick a value within a high range or low range) based on channel number if the goal is to make the output of `scan energy` more random.

Comment (bukepo) [reply examples/platforms/simulation/system.c:134]: We encountered a bug on posix platform. The bug is related to RCP with energy scan capability which causes calling platform radio APIs directly in the energy scan done callback. The bug is not easy to reproduce on a real device, so I want to add this implementation to help reproduce the bug and verify it's fix.

[the remark this answers is not in the document]
Comment (bukepo) [reply examples/platforms/simulation/system.c:134]: We also want to simulate RCP doesn't support energy scan for testing and coverage.

[the remark this answers is not in the document]
Comment (zhanglongxia) [reply examples/platforms/simulation/system.c:134]: Should we also update the test cases to cover this case in this PR?

**Comment by abtink** [comment]
@bukepo , do we still want to continue with this PR? 
Per https://github.com/openthread/openthread/pull/7393#discussion_r801147969 I think the mian intention was to be able to debug the posix and RCP timeout issue which I think we found the root cause (from https://github.com/openthread/openthread/issues/7429).

**Comment by bukepo** [comment]
> @bukepo , do we still want to continue with this PR? Per [#7393 (comment)](https://github.com/openthread/openthread/pull/7393#discussion_r801147969) I think the mian intention was to be able to debug the posix and RCP timeout issue which I think we found the root cause (from #7429).

I think this still can improve the code coverage. I can add a few tests later. Thoughts?

**Review by zhanglongxia — APPROVED** [review]
LGTM

File: examples/platforms/simulation/system.c (hunk 1 above)
Comment (bukepo) [reply examples/platforms/simulation/system.c:134]: Updated to share the RSSI simulation.

**Comment by bukepo** [comment]
@abtink I've updated the code to share RSSI simulation and added a test for energy scan enabled on RCP. Could you please take another look?

File: examples/platforms/simulation/radio.c [hunk 2]
```diff
@@ -757,6 +770,25 @@ void platformRadioUpdateFdSet(fd_set *aReadFdSet, fd_set *aWriteFdSet, int *aMax
             *aMaxFd = sTxFd;
         }
     }
+
+    if (sEnergyScanning)
+    {
+        struct timeval tv  = {0, 0};
+        uint32_t       now = otPlatAlarmMilliGetNow();
+
+        if ((int32_t)(sEnergyScanEndTime - now) > 0)
```
Comment (abtink) [inline examples/platforms/simulation/radio.c:779]: The cast from `uint` to `int` is not always safe (it is technically undefined behavior in C++ and up to toolchains). We can replace it with `uint` subtraction and comparison of result with `2^31`.

File: examples/platforms/simulation/radio.c [hunk 3]
```diff
@@ -811,11 +843,16 @@ void platformRadioProcess(otInstance *aInstance, const fd_set *aReadFdSet, const
         }
     }
 #endif
-
     if (platformRadioIsTransmitPending())
     {
         radioSendMessage(aInstance);
     }
+
+    if (sEnergyScanning && ((int32_t)(otPlatAlarmMilliGetNow() - sEnergyScanEndTime)) >= 0)
```
Comment (abtink) [inline examples/platforms/simulation/radio.c:851]: Same here.

**Review by abtink — COMMENTED** [review]
Looks good overall. One suggestion below.

**Review by abtink — APPROVED** [review]
LGTM.

**Review by jwhui — APPROVED** [review]

**Pull request merged**